subsys/bluetooth/host/audio/bap.c

```diff
@@ -20,84 +20,1323 @@
-#include "capabilities.h"
 #include "pacs_internal.h"
+#include "bap_internal.h"
 
 #define BT_DBG_ENABLED IS_ENABLED(CONFIG_BT_DEBUG_BAP)
 #define LOG_MODULE_NAME bt_bap
 #include "common/log.h"
 
 #if defined(CONFIG_BT_AUDIO_UNICAST_CLIENT)
 
+#define PAC_TYPE_UNUSED 0
+
 static struct bap_pac {
-	struct bt_audio_capability cap;
-	struct bt_codec            codec;
+	uint8_t type; /* type == PAC_TYPE_UNUSED means unused */
+	uint16_t context;
+	struct bt_codec codec;
 } cache[CONFIG_BT_MAX_CONN][CONFIG_BT_BAP_PAC_COUNT];
 
-static const struct bt_uuid *snk_uuid = BT_UUID_PACS_SNK;
-static const struct bt_uuid *src_uuid = BT_UUID_PACS_SRC;
-static const struct bt_uuid *pacs_context_uuid = BT_UUID_PACS_SUPPORTED_CONTEXT;
-static const struct bt_uuid *ase_snk_uuid = BT_UUID_ASCS_ASE_SNK;
-static const struct bt_uuid *ase_src_uuid = BT_UUID_ASCS_ASE_SRC;
-static const struct bt_uuid *cp_uuid = BT_UUID_ASCS_ASE_CP;
+static const struct bt_uuid *snk_uuid = BT_UUID_PACS_SNK;
+static const struct bt_uuid *src_uuid = BT_UUID_PACS_SRC;
+static const struct bt_uuid *pacs_context_uuid = BT_UUID_PACS_SUPPORTED_CONTEXT;
+static const struct bt_uuid *ase_snk_uuid = BT_UUID_ASCS_ASE_SNK;
+static const struct bt_uuid *ase_src_uuid = BT_UUID_ASCS_ASE_SRC;
+static const struct bt_uuid *cp_uuid = BT_UUID_ASCS_ASE_CP;
+
+
+#define EP_ISO(_iso) CONTAINER_OF(_iso, struct bt_audio_ep, iso)
+
+static struct bt_audio_ep snks[CONFIG_BT_MAX_CONN][SNK_SIZE];
+static struct bt_audio_ep srcs[CONFIG_BT_MAX_CONN][SRC_SIZE];
+
+static struct bt_gatt_subscribe_params cp_subscribe[CONFIG_BT_MAX_CONN];
+static struct bt_gatt_discover_params cp_disc[CONFIG_BT_MAX_CONN];
+
+
+/* TODO: Move the functions to avoid these prototypes */
+static int bap_ep_set_metadata(struct bt_audio_ep *ep,
+			       struct net_buf_simple *buf,
+			       uint8_t len, struct bt_codec *codec);
+
+static int bap_ep_set_codec(struct bt_audio_ep *ep, uint8_t id, uint16_t cid,
+			    uint16_t vid, struct net_buf_simple *buf,
+			    uint8_t len, struct bt_codec *codec);
+
+static void bap_ep_iso_recv(struct bt_iso_chan *chan,
+			    const struct bt_iso_recv_info *info,
+			    struct net_buf *buf)
+{
+	struct bt_audio_ep *ep = EP_ISO(chan);
+	struct bt_audio_stream_ops *ops = ep->stream->ops;
+
+	BT_DBG("stream %p ep %p len %zu", chan, ep, net_buf_frags_len(buf));
+
+	if (ops != NULL && ops->recv != NULL) {
+		ops->recv(ep->stream, buf);
+	}
+}
+
+static void bap_ep_iso_connected(struct bt_iso_chan *chan)
+{
+	struct bt_audio_ep *ep = EP_ISO(chan);
+	struct bt_audio_stream_ops *ops = ep->stream->ops;
+
+	BT_DBG("stream %p ep %p type %u", chan, ep, ep != NULL ? ep->type : 0);
+
+	if (ops != NULL && ops->connected != NULL) {
+		ops->connected(ep->stream);
+	}
+
+	if (ep->status.state != BT_AUDIO_EP_STATE_ENABLING) {
+		BT_DBG("endpoint not in enabling state: %s",
+		       bt_audio_ep_state_str(ep->status.state));
+		return;
+	}
+
+	bt_bap_ep_set_state(ep, BT_AUDIO_EP_STATE_STREAMING);
```

Comment:
This transition is only valid on the peripheral side - and only if the audio sub-system is done configuring the codec and audio data path.
On the central side the transition shall be triggered by Receiver_Start_Ready.